app/lib/analyzer/backend.dart

```diff
@@ -0,0 +1,158 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math';
+
+import 'package:gcloud/db.dart';
+
+import '../shared/utils.dart';
+
+import 'models.dart';
+
+final Random _random = new Random.secure();
+
+/// Datastore-related access methods for the analyzer service
+class AnalysisBackend {
+  final DatastoreDB db;
+
+  AnalysisBackend(this.db);
+
+  /// Gets the [Analysis] object from the Datastore. [version] and [analysis] is
+  /// optional, when they are missing the analysis from the latest rollup will
+  /// be returned.
+  Future<Analysis> getAnalysis(String package,
+      [String version, int analysis]) async {
+    final Key packageKey = db.emptyKey.append(PackageAnalysis, id: package);
+
+    if (version == null) {
+      final list = await db.lookup([packageKey]);
+      final PackageAnalysis pa = list[0];
+      if (pa == null) return null;
+      version = pa.latestVersion;
+      analysis = pa.latestAnalysis;
+    }
+
+    // version was set
+    final Key versionKey =
+        packageKey.append(PackageVersionAnalysis, id: version);
+
+    if (analysis == null) {
+      final list = await db.lookup([versionKey]);
+      final PackageVersionAnalysis pva = list[0];
+      if (pva == null) return null;
+      analysis = pva.latestAnalysis;
+    }
+
+    // analysis was set
+    final Key analysisKey = versionKey.append(Analysis, id: analysis);
+    final List list = await db.lookup([analysisKey]);
+    return list[0];
+  }
+
+  /// Create a new [Analysis] object with the current timestamp.
+  Analysis prepareAnalysis(String package, String version) {
+    final DateTime now = new DateTime.now().toUtc();
+    final Key versionKey = db.emptyKey
+        .append(PackageAnalysis, id: package)
+        .append(PackageVersionAnalysis, id: version);
+    return new Analysis()
+      ..parentKey = versionKey
+      ..packageName = package
+      ..packageVersion = version
+      ..timestamp = now;
+  }
+
+  /// Stores the analysis, and either creates or updates its parent
+  /// [PackageAnalysis] and [PackageVersionAnalysis] rollups records.
+  Future storeAnalysis(Analysis analysis) async {
+    assert(analysis.id == null);
+    await db.commit(inserts: [analysis]);
+    assert(analysis.id != null);
+
+    // update package and version too
+    await db.withTransaction((Transaction tx) async {
+      final Key packageKey =
+          db.emptyKey.append(PackageAnalysis, id: analysis.packageName);
+      final Key packageVersionKey = packageKey.append(PackageVersionAnalysis,
+          id: analysis.packageVersion);
+      final List parents = await tx.lookup([packageKey, packageVersionKey]);
+      final PackageAnalysis package = parents[0];
+      final PackageVersionAnalysis version = parents[1];
+
+      final List<Model> inserts = [];
+      if (version == null) {
+        inserts.add(_createVersion(analysis));
+      } else {
+        if (_updateVersionLatest(version, analysis)) {
```

Comment:
Maybe in another CL: In the normal pub code you added methods to the models for this IIRC, maybe we should do that here as well.